### PR TITLE
Tests for queue and stack updated

### LINKED ISSUE #154: Multiple tests written under single test name.
#### Description of the problem
generally one module has one test which has all of the tests implemented in it.

#### Example of the problem
```
def test_Queue():

    q1 = Queue(implementation='array', items=[0])
    q1.append(1)
    q1.append(2)
    q1.append(3)

    q1 = Queue(implementation='linked_list')
    q1.append(1)
    assert raises(TypeError, lambda: Queue(implementation='linked_list', items={0, 1}))
    q1 = Queue(implementation='linked_list', items = [0, 1])
    q1.append(2)
    q1.append(3)
```
This can be changed to
```
test_Queue():
   <tests for Queue abstract class>

test_ArrayQueue():
   <tests for ArrayQueue class>

test_LinkedListQueue():
    <tests for LinkedListQueue class>
```
This will increase readability, and maintaining tests would be easier.

### PR DESCRIPTION
**Pull request by HarsheetKakar**
#### References to other Issues or PRs or Relevant literature
Fixes #154 

#### Other comments
please check if the changes in __init__.py are acceptable

### CODE REVIEW

File: pydatastructs/miscellaneous_data_structures/__init__.py [hunk 1]
```diff
@@ -12,10 +12,14 @@
 
 from .stack import (
     Stack,
+    ArrayStack,
+    LinkedListStack
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/__init__.py]:
```suggestion
```
Implementation classes shouldn't be in the global namespace.

File: pydatastructs/miscellaneous_data_structures/__init__.py [hunk 2]
```diff
@@ -12,10 +12,14 @@
...
 )
 __all__.extend(stack.__all__)
 
 from .queue import (
     Queue,
+    ArrayQueue,
+    LinkedListQueue
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/__init__.py]:
```suggestion
```
+1

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 3]
```diff
@@ -3,7 +3,9 @@
 from copy import deepcopy as dc
 
 __all__ = [
-    'Queue'
+    'Queue',
+    'ArrayQueue',
+    'LinkedListQueue'
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
```
+1

File: pydatastructs/miscellaneous_data_structures/__init__.py (hunk 1 above)
Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/__init__.py]: ok let me check if its giving any errors or not

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 4]
```diff
@@ -73,6 +75,8 @@ class ArrayQueue(Queue):
 
     def __new__(cls, items=None, dtype=NoneType):
         if items is None:
+            #TODO: Better implementation of ArrayQueue(items=None, dtype=NoneType)
+            dtype = int
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
```

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 5]
```diff
@@ -3,7 +3,9 @@
 from copy import deepcopy as dc
 
 __all__ = [
-    'Stack'
+    'Stack',
+    'ArrayStack',
+    'LinkedListStack'
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
```

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 6]
```diff
@@ -82,6 +84,8 @@ class ArrayStack(Stack):
 
     def __new__(cls, items=None, dtype=NoneType):
         if items is None:
+            #TODO: Better implementation of ArrayStack(items=None, dtype=NoneType)
+            dtype = int
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
```

**Comment by czgdp1807** [comment]
Test both the implementations but use the interface classes `Queue` and `Stack` to test. Implementation classes shouldn't be accessed.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py [hunk 7]
```diff
@@ -1,9 +1,18 @@
-from pydatastructs.miscellaneous_data_structures import Queue
+from pydatastructs.miscellaneous_data_structures import (
+    Queue, ArrayQueue, LinkedListQueue)
 from pydatastructs.utils.raises_util import raises
 
 def test_Queue():
+    q1 = Queue(implementation='array', items=[0])
+    q1 = Queue()
+    q1 = Queue(implementation='linked_list')
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: Remove this function.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py [hunk 8]
```diff
@@ -1,9 +1,18 @@
...
 
+def test_ArrayQueue():
+    q = ArrayQueue()
+    q1 = Queue()
+    raises(IndexError, lambda: q1.popleft())
     q1 = Queue(implementation='array', items=[0])
+    assert type(q) == type(q1)
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: Use `_check_type`.

Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]:
```suggestion
    assert _check_type(q1, ArrayQueue)
```

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py (hunk 7 above)
Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: this function checks for implementations of the abstract datatype. If changes are made which are supposed to be common in implementations the other developer can just make changes here.

Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: Well this function has no `assert` so nothing is checked here.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py (hunk 8 above)
Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: both `_check_type(q1, Queue)` and `_check_type(q1, ArrayQueue)` will give True. The main thinking behind the assertion was to check if the object is of type Queue or of type ArrayQueue.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py (hunk 7 above)
Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: sure will take care of that.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py [hunk 9]
```diff
@@ -1,8 +1,18 @@
...
 from pydatastructs.utils.raises_util import raises
+from pydatastructs.utils.misc_util import _check_type
 
 def test_Queue():
+    q = Queue(implementation='array')
+    q1 = Queue()
+    assert type(q) is type(q1)
+    q2 = Queue(implementation='linked_list')
+    assert _check_type(q2,Queue) is True
+    assert raises(NotImplementedError, lambda: Queue(implementation=''))
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_queue.py]:
```suggestion
    q = Queue(implementation='array')
    assert _check_type(q, Queue) is True
    q2 = Queue(implementation='linked_list')
    assert _check_type(q2, Queue) is True
    assert raises(NotImplementedError, lambda: Queue(implementation=''))
```

Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: `assert type(q) is type(q1)` is to check if default implementation of Queue is ArrayQueue and not some other implementation.

File: pydatastructs/miscellaneous_data_structures/tests/test_stack.py [hunk 10]
```diff
@@ -1,10 +1,18 @@
...
 from pydatastructs.utils.raises_util import raises
+from pydatastructs.utils.misc_util import _check_type
 
 def test_Stack():
+    s = Stack(implementation='array')
+    s1 = Stack()
+    assert type(s) is type(s1)
+    s2 = Stack(implementation='linked_list')
+    assert _check_type(s2,Stack) is True
+    assert raises(NotImplementedError, lambda: Stack(implementation=''))
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_stack.py]:
```suggestion
    s = Stack(implementation='array')
    assert _check_type(s, ArrayStack) is True
    s2 = Stack(implementation='linked_list')
    assert _check_type(s2, LinkedListStack) is True
    assert raises(NotImplementedError, lambda: Stack(implementation=''))
```

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py (hunk 9 above)
Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: `assert _check_type(q, ArrayQueue) is True`
Let me try this one.

File: pydatastructs/miscellaneous_data_structures/tests/test_queue.py (hunk 8 above)
Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/tests/test_queue.py]: So basically it is to be verified that correct implementation queue is created. Well then for now we can go for using `type` for now for this case.

File: pydatastructs/miscellaneous_data_structures/tests/test_stack.py [hunk 11]
```diff
@@ -1,10 +1,18 @@
...
 from pydatastructs.linear_data_structures import OneDimensionalArray
 from pydatastructs.utils.raises_util import raises
+from pydatastructs.utils.misc_util import _check_type
 
 def test_Stack():
+    s = Stack(implementation='array')
+    s1 = Stack()
+    assert type(s) is type(s1)
+    s2 = Stack(implementation='linked_list')
+    assert _check_type(s2,Stack) is True
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_stack.py]:
```suggestion
    assert _check_type(s2, Stack) is True
```

Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/tests/test_stack.py]: How will we check if that `s2` is `LinkedListStack`?
Probably something like,
`_check_type(s2, Stack) and _check_type(s2, LinkedListStack)` should work well.

Comment (HarsheetKakar) [reply pydatastructs/miscellaneous_data_structures/tests/test_stack.py]: I have made some changes please review it again.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 12]
```diff
@@ -3,7 +3,7 @@
 from copy import deepcopy as dc
 
 __all__ = [
-    'Stack'
+    'Stack',
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
    'Stack'
```

**Pull request merged**